src/core/Akka/Actor/Message.cs

```diff
@@ -5,26 +5,43 @@
 // </copyright>
 //-----------------------------------------------------------------------
 
+using System;
+
 namespace Akka.Actor
 {
-    //Note: this is a struct in order to lower GC pressure, it will be removed once the mailbox Run call goes out of scope. //Roger
-
     /// <summary>
-    ///     Envelope class, represents a message and the sender of the message.    
+    /// Envelope class, represents a message and the sender of the message.    
     /// </summary>
     public struct Envelope
     {
+        public Envelope(object message, IActorRef sender, ActorSystem system)
+        {
+            if (message == null)
+            {
+                throw new ArgumentNullException(nameof(message));
+            }
+
+            Message = message;
+            Sender = sender != ActorRefs.NoSender ? sender : system.DeadLetters;
+        }
+
+        public Envelope(object message, IActorRef sender)
```

Comment:
I should have made this constructor as internal, but I did not manage to do it. I've got some conflicts in Persistence